src/index.ts

```diff
@@ -0,0 +1,2719 @@
+    if (diff > 0) {
+      let subtractlayerslater = 0;
+      //Ironically, this edge case would be unnecessary if we had 'negative layers'.
+      while (Number.isFinite(result.mag) && result.mag < 10) {
+        result.mag = Math.pow(10, result.mag);
+        ++subtractlayerslater;
+      }
+
+      //A^(10^B) === C, solve for B
+      //B === log10(logA(C))
+
+      if (result.mag > 1e10) {
+        result.mag = Math.log10(result.mag);
+        result.layer++;
+      }
+
+      //Note that every integer slog10 value, the formula changes, so if we're near such a number, we have to spend exactly enough layerdiff to hit it, and then use the new formula.
+      const diffToNextSlog = Math.log10(Math.log(1e10) / Math.log(result.mag), 10); // FIXME: This has an extra argument
+      if (diffToNextSlog < diff) {
+        result.mag = Math.log10(1e10);
+        result.layer++;
+        diff -= diffToNextSlog;
+      }
+
+      result.mag = Math.pow(result.mag, Math.pow(10, diff));
+
+      while (subtractlayerslater > 0) {
+        result.mag = Math.log10(result.mag);
+        --subtractlayerslater;
+      }
+    } else if (diff < 0) {
+      let subtractlayerslater = 0;
+
+      while (Number.isFinite(result.mag) && result.mag < 10) {
+        result.mag = Math.pow(10, result.mag);
+        ++subtractlayerslater;
+      }
+
+      if (result.mag > 1e10) {
+        result.mag = Math.log10(result.mag);
+        result.layer++;
+      }
+
+      const diffToNextSlog = Math.log10(1 / Math.log10(result.mag));
+      if (diffToNextSlog > diff) {
+        result.mag = 1e10;
+        result.layer--;
+        diff -= diffToNextSlog;
+      }
+
+      result.mag = Math.pow(result.mag, Math.pow(10, diff));
+
+      while (subtractlayerslater > 0) {
+        result.mag = Math.log10(result.mag);
+        --subtractlayerslater;
+      }
+    }
+
+    while (result.layer < 0) {
+      result.layer++;
+      result.mag = Math.log10(result.mag);
+    }
+    result.normalize();
+    return result;
+  }
+
+  //layeradd: like adding 'diff' to the number's slog(base) representation. Very similar to tetrate base 'base' and iterated log base 'base'.
+  public layeradd(diff: number, base: DecimalSource): Decimal {
+    const slogthis = this.slog(base).toNumber();
+    const slogdest = slogthis + diff;
+    if (slogdest >= 0) {
+      return Decimal.tetrate(base, slogdest);
+    } else if (!Number.isFinite(slogdest)) {
+      return Decimal.dNaN;
+    } else if (slogdest >= -1) {
+      return Decimal.log(Decimal.tetrate(base, slogdest + 1), base);
+    } else {
+      // FIXME: Is this supposed to return this value??
+      Decimal.log(Decimal.log(Decimal.tetrate(base, slogdest + 2), base), base);
+      throw "Unhandled behavior";
```

Comment:
Not sure about this one, but this seems logical since the others return a value

```suggestion
      return Decimal.log(Decimal.log(Decimal.tetrate(base, slogdest + 2), base), base);
```